app/services/list_volunteer_opportunities.rb

```diff
@@ -1,55 +1,80 @@
+# google map.  Those markers are merged with other markers generated from the
+# list of organisations in our local database that have volunteers.
+
+# This means the title of the service is somewhat misleading, although it is
+# used in the Volunteer Ops controller.  To properly follow our CRUD services
+# model this service should be titled ListOrganisationsWithVolunteers, or
+# ListOrganisations and take a parameter to indicate we want to list
+# Organisations that have associated volunteer opportunities.  It even implies
+# that the Volunteer Op view should arguably be loaded from the
+# OrganisationsController
+
+# As it stands the current code in this service is mainly concerned with
+# processing one or more pages of JSON data from the doit API and transforming
+# that into an array of OpenStructs that can be handled by the current view. The
+# code also calls back to the controller to create the necessary google map
+# markers and merge the two sets together; handling cases where one or other
+# might be missing
 class ListVolunteerOpportunities
 
-  def self.with(listener, organisations, feature = Feature, http_party = HTTParty)
-    new(listener, organisations, feature, http_party).send(:run)
+  def self.with(listener, organisations, feature = Feature,
+                http = HTTParty)
+    new(listener, organisations, feature, http).send(:run)
   end
 
   private
 
-  attr_reader :organisations, :listener, :feature, :http_party
+  attr_reader :organisations, :listener, :feature, :http
 
-  def initialize(listener, organisations, feature, http_party)
+  def initialize(listener, organisations, feature, http)
     @listener = listener
     @organisations = organisations
     @feature = feature
-    @http_party = http_party
+    @http = http
   end
 
+  HOST = 'https://api.do-it.org'
+  HREF = '/v1/opportunities?lat=51.5978&lng=-0.3370&miles=0.5'
+
   def run
     harrow_markers = listener.build_map_markers(organisations)
     return [harrow_markers, []] unless feature.active? :doit_volunteer_opportunities
-    doit_orgs = get_doit_orgs
+    doit_orgs = doit_orgs_for_multiple_pages_of_json
     doit_markers = listener.build_map_markers(doit_orgs, :doit, false)
     markers = merge_json_markers(harrow_markers, doit_markers)
     [markers, doit_orgs]
   end
 
-	def get_doit_orgs
-		host = 'https://api.do-it.org'
-    href = "/v1/opportunities\?lat\=51.5978\&lng\=-0.3370\&miles\=0.5 "
-    doit_orgs = [] 
-    while href
-      url = host + href
-      response = http_party.get(url)
-      if response.body && response.body != '[]'
-        opportunities = JSON.parse(response.body)['data']['items']
-        get_doit_orgs_from_one_page(doit_orgs, opportunities)
-        next_page = JSON.parse(response.body)['links']['next']
-      end
-      href = next_page ? next_page['href'] : nil
+  def doit_orgs_for_multiple_pages_of_json
+    href = HREF
+    doit_orgs = []
+    while href = process_doit_json_page(http.get("#{HOST}#{href}"), doit_orgs)
```

Comment:
The while means it will run multiple times until a new href value is not assigned, I.e. We have come to the last page of data.

Not entirely happy with this though. Feel like someone must already have a library that handles paginating through JSON responses